src/stratigraphy/util/geometric_line_utilities.py

```diff
@@ -223,36 +229,106 @@ def merge_parallel_lines(lines: list[Line], tol: int = 8, angle_threshold: float
         return merged_lines
 
 
+def odr_regression(x: ArrayLike, y: ArrayLike) -> tuple[float, float]:
+    """Perform orthogonal distance regression on the given data.
+
+    Solves the line equation y = mx + c for the best fit values of m and c.
+    Different to linear regression, ODR takes into account the error in both x and y.
+    As such, it can handle horizontal and vertical lines.
+
+    Note: For vertical lines, the slope is nan.
+
+    Args:
+        x (ArrayLike): The x-coordinates of the data.
+        y (ArrayLike): The y-coordinates of the data.
+
+    Returns:
+        tuple[float, float]: (m, c), the best fit values for the line equation.
+    """
+
+    # Define a function to fit the data with.
+    def linear_func(p, x):
+        m, c = p
+        return m * x + c
+
+    linear_model = odr.Model(linear_func)
+    data = odr.RealData(x, y)
+    odr_model = odr.ODR(data, linear_model, beta0=[0.0, 1.0])  # kept the initial guess as default
+
+    out = odr_model.run()
+
+    return out.beta[0], out.beta[1]
+
+
 def _merge_lines(line1: Line, line2: Line) -> Line:
-    """Merge two lines.
+    """Merge two lines into one.
+
+    The algorithm performes odr regression on the points of the two lines to find the best fit line.
+    Then, it calculates the orthogonal projection of the four points onto the best fit line and takes the two points
+    that are the furthest apart. These two points are then used to create the merged line.
 
-    Note: This method is not robust for near vertical lines. We assume
-    an ordering in the x-coordinates of the lines.
+    Note: Strictly vertical lines cannot be handled. In that case the slope is nan. A warning is thrown
+    and the first line is returned.
```

Comment:
Here's a derivation of a fitting line that also supports the vertical case: https://www.researchgate.net/publication/272179120_A_tutorial_on_the_total_least_squares_method_for_fitting_a_straight_line_and_a_plane
I did not find a Python implementation of this yet, though. Maybe you can find something? Or otherwise, it does not look too difficult to implement the actual formulas ourselves.